src/graph/graphcontext.c

```diff
@@ -0,0 +1,184 @@
+  RedisModule_FreeString(ctx, rm_graphctx);
+
+  return key;
+}
+
+GraphContext* GraphContext_New(RedisModuleCtx *ctx, RedisModuleString *rm_name, const char *graph_name) {
+  Graph *g = Graph_Get(ctx, rm_name);
+  if (!g) return NULL;
+
+  GraphContext *gc = malloc(sizeof(GraphContext));
+  gc->ctx = ctx;
+  gc->g = g;
+  gc->graph_name = strdup(graph_name);
+  gc->node_stores = calloc(GRAPH_DEFAULT_LABEL_CAP, sizeof(LabelStore*));
+  gc->relation_stores = calloc(GRAPH_DEFAULT_RELATION_CAP, sizeof(LabelStore*));
+  gc->indices = NULL;
+
+  gc->relation_cap = GRAPH_DEFAULT_RELATION_CAP;
+  gc->relation_count = 0;
+  gc->node_cap = GRAPH_DEFAULT_LABEL_CAP;
+  gc->node_count = 0;
+  gc->index_count = 0;
+
+  // LabelStore_Get will construct allstores if they don't already exist,
+  // though I'd like to refactor its logic somewhat more generally.
+  gc->node_allstore = LabelStore_Get(ctx, STORE_NODE, graph_name, NULL);
+  gc->relation_allstore = LabelStore_Get(ctx, STORE_EDGE, graph_name, NULL);
+
+  // TODO is there a way to populate store arrays here?
+
+  RedisModuleKey *key = GraphContext_Key(ctx, graph_name);
+  assert(RedisModule_ModuleTypeGetType(key) == REDISMODULE_KEYTYPE_EMPTY);
+  RedisModule_ModuleTypeSetValue(key, GraphContextRedisModuleType, gc);
+  RedisModule_CloseKey(key);
+
+  return gc;
+}
+
+GraphContext* GraphContext_Get(RedisModuleCtx *ctx, RedisModuleString *rs_graph_name, const char *graph_name) {
+  Graph *g = Graph_Get(ctx, rs_graph_name);
+  if (!g) return NULL;
+
+  RedisModuleKey *key = GraphContext_Key(ctx, graph_name);
+  assert(RedisModule_ModuleTypeGetType(key) == GraphContextRedisModuleType);
+  GraphContext *gc = RedisModule_ModuleTypeGetValue(key);
+  RedisModule_CloseKey(key);
+
+  gc->ctx = ctx;
+  gc->g = g;
+  gc->node_allstore = LabelStore_Get(ctx, STORE_NODE, graph_name, NULL);
+  gc->relation_allstore = LabelStore_Get(ctx, STORE_EDGE, graph_name, NULL);
+
+  // Retrieve all label stores?
+  // Can iterate over all label strings or serialize them within GraphContext value
+
+  return gc;
+}
+
+// TODO next two functions are nearly identical; maybe consolidate
+LabelStore* GraphContext_AddNode(GraphContext *gc, const char *label) {
+  if(gc->node_count == gc->node_cap) {
+    gc->node_cap += 4;
+    gc->node_stores = realloc(gc->node_stores, gc->node_cap * sizeof(LabelStore*));
+    memset(gc->node_stores + gc->node_count, 0, 4 * sizeof(LabelStore*));
+  }
+  // Graph_AddLabel(gc->g);
+  LabelStore *store = LabelStore_New(gc->ctx, STORE_NODE, gc->graph_name, label, gc->node_count);
+  gc->node_stores[gc->node_count] = store;
+  gc->node_count++;
+
+  return store;
+}
+
+LabelStore* GraphContext_AddRelation(GraphContext *gc, const char *label) {
+  if(gc->relation_count == gc->relation_cap) {
+    gc->relation_cap += 4;
+    gc->relation_stores = realloc(gc->relation_stores, gc->relation_cap * sizeof(LabelStore*));
+    memset(gc->relation_stores + gc->relation_count, 0, 4 * sizeof(LabelStore*));
```

Comment:
If the GraphContext was loaded from rdb, we initially won't have any stores cached - they'll all need to be retrieved from the keyspace on first access. We NULL-check the stores arrays to see if this is the first access or not.

The idea of serializing stores as part of the context rather than doing lazy lookups is growing on me, though, in which case this logic is unnecessary.